File: packages/go/qorechain/utils/validators.go
package utils

import (
	"encoding/hex"
	"strings"

	"github.com/mr-tron/base58"
)

// IsValidEVMAddress reports whether s is a structurally valid EVM address: a
// "0x"-prefixed 40-hex-character string. It does not verify the EIP-55 checksum;
// use ToChecksumAddress to produce a checksummed form.
func IsValidEVMAddress(s string) bool {
	if len(s) != 42 {
		return false
	}
	if s[0] != '0' || (s[1] != 'x' && s[1] != 'X') {
		return false
	}
	_, err := hex.DecodeString(s[2:])
	return err == nil
}

// IsValidSVMAddress reports whether s is a structurally valid SVM address: a
// base58 string that decodes to exactly 32 bytes.
func IsValidSVMAddress(s string) bool {
	decoded, err := base58.Decode(s)
	if err != nil {
		return false
	}
	return len(decoded) == 32
}

// ToChecksumAddress returns the EIP-55 mixed-case checksum form of an EVM
// address. It returns an error if s is not a valid 20-byte hex address.
func ToChecksumAddress(s string) (string, error) {
	if !IsValidEVMAddress(s) {
		return "", errInvalidEVMAddress(s)
	}
	lower := strings.ToLower(s[2:])
	hash := Keccak256([]byte(lower))
	out := []byte("0x")
	for i := 0; i < len(lower); i++ {
		c := lower[i]
		if c >= '0' && c <= '9' {
			out = append(out, c)
			continue
		}
		// Uppercase the hex letter when the corresponding hash nibble >= 8.
		nibble := hash[i/2]
		if i%2 == 0 {
			nibble >>= 4
		}
		nibble &= 0x0f
		if nibble >= 8 {
			out = append(out, byte(strings.ToUpper(string(c))[0]))
		} else {
			out = append(out, c)
		}
	}
	return string(out), nil
}

type errInvalidEVMAddress string

func (e errInvalidEVMAddress) Error() string {
	return "invalid EVM address: " + string(e)
}

File: packages/java/src/main/java/io/github/qorechain/address/Validators.java
package io.github.qorechain.address;

import io.github.qorechain.utils.Base58;
import io.github.qorechain.utils.Hashing;
import java.nio.charset.StandardCharsets;
import org.bouncycastle.math.ec.rfc8032.Ed25519;

/**
 * Cross-VM address validation and the EIP-55 mixed-case checksum.
 *
 * <p>QoreChain spans three address formats: bech32 ({@code qor1...}, in
 * {@link Address}), EVM hex ({@code 0x...}, 20 bytes), and SVM ed25519 public
 * keys (base58, 32 bytes). These helpers validate the EVM and SVM forms and
 * expose the EIP-55 checksum used by EVM tooling.
 */
public final class Validators {

    private Validators() {}

    /** Test whether {@code address} is {@code 0x} + exactly 40 hex chars. Case-insensitive. */
    public static boolean isValidEvmAddress(String address) {
        return address.matches("0x[0-9a-fA-F]{40}");
    }

    /**
     * Compute the EIP-55 mixed-case checksum form of an EVM address.
     *
     * @throws IllegalArgumentException if not a valid 20-byte hex address.
     */
    public static String toChecksumAddress(String address) {
        String body =
                (address.startsWith("0x") || address.startsWith("0X")
                                ? address.substring(2)
                                : address)
                        .toLowerCase();
        if (!body.matches("[0-9a-f]{40}")) {
            throw new IllegalArgumentException("invalid EVM address: " + address);
        }
        byte[] hash = Hashing.keccak256(body.getBytes(StandardCharsets.US_ASCII));
        StringBuilder out = new StringBuilder("0x");
        for (int i = 0; i < body.length(); i++) {
            char ch = body.charAt(i);
            if (ch >= '0' && ch <= '9') {
                out.append(ch);
            } else {
                int hashByte = hash[i >> 1] & 0xff;
                int nibble = (i % 2 == 0) ? (hashByte >> 4) : (hashByte & 0x0f);
                out.append(nibble >= 8 ? Character.toUpperCase(ch) : ch);
            }
        }
        return out.toString();
    }

    /**
     * Test whether {@code address} is a correctly EIP-55-checksummed EVM address.
     * All-lowercase / all-uppercase addresses (no checksum info) return false.
     * Never throws.
     */
    public static boolean isChecksumAddress(String address) {
        if (!isValidEvmAddress(address)) {
            return false;
        }
        String body = address.substring(2);
        if (body.equals(body.toLowerCase()) || body.equals(body.toUpperCase())) {
            return false;
        }
        try {
            return toChecksumAddress(address).equals(address);
        } catch (RuntimeException e) {
            return false;
        }
    }

    /**
     * Test whether {@code address} is a valid SVM public key: a base58 string
     * decoding to exactly 32 bytes that is a valid ed25519 curve point. Off-curve
     * PDAs are rejected. Never throws.
     */
    public static boolean isValidSvmAddress(String address) {
        byte[] bytes;
        try {
            bytes = Base58.decode(address);
        } catch (RuntimeException e) {
            return false;
        }
        if (bytes.length != 32) {
            return false;
        }
        return Ed25519.validatePublicKeyFull(bytes, 0);
    }
}
